lib/flutter/flutter_daemon.dart

```diff
@@ -384,6 +408,117 @@ class AppDomain extends Domain {
       (Map<String, dynamic> app) => new DiscoveredApp(app['id'], app['observatoryDevicePort'])
     ));
   }
+
+  DaemonApp createDaemonApp(String appId) => new DaemonApp(this, appId);
+}
+
+class DaemonApp {
+  final AppDomain daemon;
+  final String appId;
+
+  StreamSubscriptions _subs = new StreamSubscriptions();
+  Completer _stoppedCompleter = new Completer();
+  Completer<DebugPortAppEvent> _debugPortCompleter = new Completer<DebugPortAppEvent>();
+  StreamController<LogAppEvent> _logController = new StreamController<LogAppEvent>.broadcast();
+
+  DaemonApp(this.daemon, this.appId) {
+    // listen for the debugPort event
+    _subs.add(daemon.onAppDebugPort
+        .where((AppEvent event) => event.appId == appId)
+        .listen((DebugPortAppEvent event) {
+      if (!_debugPortCompleter.isCompleted) {
+        _debugPortCompleter.complete(event);
+      }
+    }));
+
+    // listen for logs
+    _subs.add(daemon.onAppLog.where((AppEvent event) => event.appId == appId).listen((LogAppEvent event) {
+      _logController.add(event);
+    }));
+
+    // listen for app termination
+    _subs.add(daemon.onAppStop.where((AppEvent event) => event.appId == appId).listen((_) {
+      _dispose();
+    }));
+  }
+
+  Future<bool> restart(String appId) => daemon.restart(appId);
+
+  Future<bool> stop(String appId) {
+    return daemon.stop(appId).timeout(
+      new Duration(seconds: 2),
+      onTimeout: () {
+        _dispose();
```

Comment:
Hmm, if we get the `onAppStop`, then _dispose will be called. The timeout here is a guard against the daemon process not being able to stop the app for some reason. It should be rare, but in the event the stop does fail, this timeout will let us remove the app launch from the UI.